### DISCUSSION TITLE
Roslyn scripting: Access classes from the project

### DISCUSSION THREAD

**Discussion by DerEinherjer** [dotnet/roslyn#50461] [Q&A] [answered]
I want to use roslyn to interact with the running project, like a console. At the moment I wrote to classes which are connected over a socktstream. The client class reads the normal console (, packages the text) and sends it to the server class. There I execute it either with  
```
if(scriptState != null)
    scriptState = await scriptState.ContinueWithAsync((string)msg.Object, ScriptOptions.Default.WithImports("Authentication.Certificates"));
else
    scriptState = await CSharpScript.RunAsync((string)msg.Object, ScriptOptions.Default.WithImports("Authentication.Certificates"));
```
That's all I have at the moment. Now I need to get Roslyn to know all relevant classes/objects to interact with the project. There is no more Roslyn related code at the moment.
WithImports() and WithReferences() didn't work, I still got "error CS0246".
I already ask that on the C# discord, they think I have to put the needed classes into a .dll. But if I am not wrong I'm able to give any object as global variable to Roslyn. Wouldn't that mean that Roslyn does know the classes in that project? If that's true there should be a more easy way to have access to those classes from within Roslyn.
So what's the best way to do that?

**Comment by tmat** [comment]
If you pass in a global object (like in this [sample](https://github.com/dotnet/roslyn/blob/master/docs/wiki/Scripting-API-Samples.md#parameterize-a-script)), the assembly containing the type of the global object and its dependencies are automatically referenced.

**Comment by tmat** [comment]
> WithReferences() didn't work, I still got "error CS0246".

If you are seeing errors please include the full repro.

**Reply by DerEinherjer** [reply]
With the above mentioned code i executed "using Authentication.Certificates;" and "new X509CertificateUtils();" both resulted in the following error:
```
(1,5): error CS0246: Der Typ- oder Namespacename "Authentication" wurde nicht gefunden (möglicherweise fehlt eine using-Direktive oder ein Assemblyverweis).
The type- or namespace "Authentication" was not found (maybe a using directive or assembly is missing).
```
I added a global variable of type "X509CertificateUtils" and could use it, but still was not able to create a new object of this type. Also the call to a static function of this class resulted either in "(1,1): error CS0103: Der Name "X509CertificateUtils" ist im aktuellen Kontext nicht vorhanden."(Class bot found in current context) or "(1,1): error CS0176: Auf den Member "X509CertificateUtils.CreateRootCertificat(string)" kann nicht mit einem Instanzverweis zugegriffen werden. Qualifizieren Sie ihn stattdessen mit einem Typnamen."(Don'T call a static function over a object).
The added global variable looked like this:
public class Global
```
{
    public int DEBUG_INT = 1;
    public double DEBUG_DOUBLE = 0.1;
    public X509CertificateUtils certUtils = new X509CertificateUtils();
}
```
and
`scriptState = await CSharpScript.RunAsync((string)msg.Object, ScriptOptions.Default.WithImports("Authentication.Certificates"), globals:this.global);`

**Comment by tmat** [comment, marked as answer]
What assembly is `X509CertificateUtils` defined in?

**Reply by DerEinherjer** [reply]
The assembly of the project from which I call roslyn.

**Reply by AraHaan** [reply]
You would have to pass the object in the program as a global object instance in the script state.

Also you would have to make the entire class (object) public too for roslyn scripting to pick it up as well.

**Reply by DerEinherjer** [reply]
I'm aware of this option, but that would not allow for the use of static function, which I would like to have for the utility functions in X509CertificateUtils. And it does not allow for the creation of objects within roslyn (which I don't need at the moment, but would be nice to have).

**Reply by AraHaan** [reply]
I would recommend outputting those utils to an dll then and have the scripts use the #r directive to reference the dll that would contain those utils so you could use tose static methods then.

**Reply by DerEinherjer** [reply]
As I can't put a list of object in the global and cast it back to what it was within roslyn aswell, it seems my only option is to us .dlls. Thank you both for your help.
